core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java

```diff
@@ -0,0 +1,189 @@
+     *
+     * @param obj           object to be serialized
+     * @param blobContainer blob container
+     * @param name          blob name
+     * @throws IOException
+     */
+    public void writeAtomic(T obj, BlobContainer blobContainer, String name) throws IOException {
+        String blobName = blobName(name);
+        String tempBlobName = tempBlobName(name);
+        writeBlob(obj, blobContainer, tempBlobName);
+        blobContainer.move(tempBlobName, blobName);
+    }
+
+    /**
+     * Writes properly formatted blob in atomic manner.
+     * <p/>
+     * The blob will be compressed and checksum will be written if required.
+     *
+     * @param obj           object to be serialized
+     * @param blobContainer blob container
+     * @param name          blob name
+     * @throws IOException
+     */
+    public void write(T obj, BlobContainer blobContainer, String name) throws IOException {
+        String blobName = blobName(name);
+        writeBlob(obj, blobContainer, blobName);
+    }
+
+    public void writeBlob(T obj, BlobContainer blobContainer, String blobName) throws IOException {
+        BytesReference bytes = write(obj);
+        try (OutputStream outputStream = blobContainer.createOutput(blobName)) {
+            final String resourceDesc = "ChecksumBlobStoreFormat.writeBlob(blob=\"" + blobName + "\")";
+            try (OutputStreamIndexOutput indexOutput = new OutputStreamIndexOutput(resourceDesc, outputStream, BUFFER_SIZE)) {
+                CodecUtil.writeHeader(indexOutput, codec, VERSION);
+                try (OutputStream indexOutputOutputStream = new org.elasticsearch.common.lucene.store.OutputStreamIndexOutput(indexOutput) {
+                    @Override
+                    public void close() throws IOException {
+                        // this is important since some of the XContentBuilders write bytes on close.
+                        // in order to write the footer we need to prevent closing the actual index input.
+                    } }) {
+                    bytes.writeTo(indexOutputOutputStream);
+                }
+                CodecUtil.writeFooter(indexOutput);
+            }
+        }
+    }
+
+    protected BytesReference write(T obj) throws IOException {
+        try (BytesStreamOutput bytesStreamOutput = new BytesStreamOutput()) {
+            if (compress) {
+                try (StreamOutput compressedStreamOutput = CompressorFactory.defaultCompressor().streamOutput(bytesStreamOutput)) {
+                    write(obj, compressedStreamOutput);
+                }
+            } else {
+                write(obj, bytesStreamOutput);
+            }
+            return bytesStreamOutput.bytes();
+        }
+    }
+
+    protected void write(T obj, StreamOutput streamOutput) throws IOException {
+        try (XContentBuilder builder = XContentFactory.contentBuilder(xContentType, streamOutput)) {
+            if (wrapper) {
```

Comment:
Maybe we can add a comment here (or where `wrapper` is defined) about why it's necessary?